[shard] Sharder and ShardingPlan prototype

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73873



Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model, this including `plan` for sharding the parameters, and `output_plan` for tagging the output layout, `return_local_tensor` for converting back to DDP.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.

TODO:
next PR we will introduce Extensible Sharder and ShardingPlanner.

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

Commit:
[shard] Sharder and ShardingPlan prototype

Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.
3. We also introduce a extensible ShardingPlanner and this extensible ShardingPlanner can be used by advanced users to build their own ShardingPlan easily.

TODO:
* implement shard_module_swapper
* implement reshard_output

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

[ghstack-poisoned]

fduwjj commented on torch/distributed/_shard/sharding_plan/api.py:
```diff
@@ -0,0 +1,25 @@
+class ShardingPlan(object):
+    """
+    Representation of a sharding plan.
+    Dict keyed by FQN of parameter to PlacementSpec.
```
N00b question, what's a FQN?

fduwjj commented on torch/distributed/_shard/api.py:
```diff
@@ -118,3 +123,80 @@ def shard_parameter(
+        param_keys = list(mod._parameters.keys())
+        for k in param_keys:
+            v = mod._parameters[k]
+            name = mod_prefix + ('.' if mod_prefix else '') + k
```
N00b question, is this expected? `mod_prefix.mod_prefix`? But why we don't check `mod_prefix` for the first one?

wanchaol replied:
can you elaborate more about what's `mod_prefix.mod_prefix`? we are just concating a module prefix together with its parameter name.

fduwjj commented on test/distributed/_shard/sharding_plan/test_sharding_plan.py:
```diff
@@ -0,0 +1,149 @@
+    sys.exit(0)
+
+
+class SimpleMegatronLM(nn.Module):
+    def __init__(self, linear_size, rank=None):
+        super().__init__()
+        self.fc1 = nn.Linear(*linear_size[0])
+        self.gelu = nn.GELU()
+        self.fc2 = nn.Linear(*linear_size[1])
+        if rank:
+            self.fc1.cuda(rank)
+            self.fc2.cuda(rank)
+
+    def forward(self, inp):
+        return self.fc2(self.gelu(self.fc1(inp)))
```
Maybe move this to util or `test_common`? My guess is that we might need this very often down the road.

fduwjj commented on torch/distributed/_shard/sharding_plan/api.py:
```diff
@@ -0,0 +1,25 @@
+    Default ShardingPlanner interface, can be extended and
+    implement advanced sharding strategies.
+    """
+    @abc.abstractmethod
```
Is it possible that we give users an option to not shard some part of the module?

wanchaol replied:
yeah, user could manually pass a ShardingPlan that only shards some certain parameters.

fduwjj reviewed (approved):
Great work, thanks! Overall, this looks good to me. Just some nits and questions for learning purpose.

pritamdamania87 commented on test/distributed/_shard/sharding_plan/test_sharding_plan.py:
```diff
@@ -0,0 +1,149 @@
+            )
+
+            # shard the module with the provided sharding plan
+            shard_module(megatron_lm, sharding_plan)
```
Don't we need to `reshard` back to dataparallel after fc2? We need a way to specify that as well.

wanchaol replied:
Yep just updated the PR to support `reshard_output` + `collect_local_shards`, all documented in the `ShardingPlan` class, let me know if this make sense.

pritamdamania87 commented on test/distributed/_shard/sharding_plan/test_sharding_plan.py:
```diff
@@ -0,0 +1,149 @@
+
+            # make sure we can run sharded computation
+            input = torch.rand(22, 17).cuda(self.rank)
+            sharded_output = megatron_lm(input)
```
We should validate this output by running a local version and comparing against it.

pritamdamania87 commented on torch/distributed/_shard/api.py:
```diff
@@ -118,3 +123,80 @@ def shard_parameter(
+        Implementation. Should return ``None`` if no swapping should be
+        performed.
+
+        The passed in module would consist of ShardedTensors and a common way
```
Maybe elaborate a bit more in terms of where the ShardedTensors come from. The sharder would produce ShardedTensors for modules based on the ShardingPlan and then call the ShardedModuleSwapper.

pritamdamania87 commented on torch/distributed/_shard/api.py:
```diff
@@ -118,3 +123,80 @@ def shard_parameter(
+
+    Args:
+        module: The module to apply sharding to
+        sharding_plan: Dict from param name to PlacementSpec/ShardingSpec
```
This is not a Dict but a `ShardingPlan` type? Shouldn't we include types in parenthesis as part of the dos?

pritamdamania87 commented on torch/distributed/_shard/api.py:
```diff
@@ -118,3 +123,80 @@ def shard_parameter(
+        module: The module to apply sharding to
+        sharding_plan: Dict from param name to PlacementSpec/ShardingSpec
+        to apply to each parameter.
+        sharded_module_swapper: Implementation of ShardedModuleSwapper for module
```
We should mention these parameters are optional.

pritamdamania87 commented on torch/distributed/_shard/api.py:
```diff
@@ -118,3 +123,80 @@ def shard_parameter(
+    plan: ShardingPlan,
+    sharded_module_swapper: ShardedModuleSwapper = None,
+    reshard_output: ShardingPlan = None,
+    src_rank=0,
```
`src_rank` is not documented.

pritamdamania87 commented on torch/distributed/_shard/sharding_plan/api.py:
```diff
@@ -0,0 +1,25 @@
+    Representation of a sharding plan.
+    Dict keyed by FQN of parameter to PlacementSpec.
+    """
+    plan: Dict[str, ShardingSpec]
```
This should be PlacementSpec?

wanchaol replied:
I decoupled PlacementSpec with ShardingSpec in the extensible ShardingSpec PR as I don't exactly know how the PlacementSpec base class should behave, maybe a `PlacementSpec` could just be an empty class and ShardingSpec be a subclass of it? (in our proposal a PlacementSpec contains a `apply()-> Tensor` abstract method, which is not applicable to ShardingSpec I think.

pritamdamania87 replied:
I think it should be PlacementSpec and we probably need to add a unit test for that too. The reason being that one type of sharding is just placing a Tensor a on gpu 0 and Tensor b on gpu 1 etc.

I think there is a DevicePlacementSpec subclass of PlacementSpec whose `apply` method can essentially just move the Tensor to the appropriate device.

pritamdamania87 commented on torch/distributed/_shard/sharding_plan/api.py:
```diff
@@ -0,0 +1,25 @@
+    implement advanced sharding strategies.
+    """
+    @abc.abstractmethod
+    def build_plan(self, module: nn.Module) -> ShardingPlan:
```
Add docs for this?

pritamdamania87 reviewed (commented):
Awesome work, the overall structure looks great!

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"


Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.
3. We also introduce a extensible ShardingPlanner and this extensible ShardingPlanner can be used by advanced users to build their own ShardingPlan easily.

TODO:
* implement shard_module_swapper
* implement reshard_output

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

[ghstack-poisoned]

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"




Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.
3. We also introduce a extensible ShardingPlanner and this extensible ShardingPlanner can be used by advanced users to build their own ShardingPlan easily.

TODO:
* implement shard_module_swapper

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

[ghstack-poisoned]

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"




Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.
3. We also introduce a extensible ShardingPlanner and this extensible ShardingPlanner can be used by advanced users to build their own ShardingPlan easily.

TODO:
* implement shard_module_swapper

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

[ghstack-poisoned]

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"




Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.
3. We also introduce a extensible ShardingPlanner and this extensible ShardingPlanner can be used by advanced users to build their own ShardingPlan easily.

TODO:
* implement shard_module_swapper

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

[ghstack-poisoned]

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"




Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.
3. We also introduce a extensible ShardingPlanner and this extensible ShardingPlanner can be used by advanced users to build their own ShardingPlan easily.

TODO:
* implement shard_module_swapper

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

[ghstack-poisoned]

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"




Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.
3. We also introduce a extensible ShardingPlanner and this extensible ShardingPlanner can be used by advanced users to build their own ShardingPlan easily.

TODO:
* implement shard_module_swapper

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharding_plan/api.py:
```diff
@@ -0,0 +1,63 @@
+            keyed by the name of module to ShardingSpec("" in key means the root module).
+            If specified, outputs are resharded according to the provided sharding specs.
+            Default: `None`
+        collect_local_shards (List[str], optional): a list of string, each element enables
```
Should we mention `""` key for this too?

pritamdamania87 commented on torch/distributed/_shard/sharding_plan/api.py:
```diff
@@ -0,0 +1,63 @@
+            If specified, outputs are resharded according to the provided sharding specs.
+            Default: `None`
+        collect_local_shards (List[str], optional): a list of string, each element enables
+            a module's sharded output to be gathered as a Tensor to ensure further processsing
```
Let's add more details about how we gather this as a Tensor. For example, it might seem to use we actually do comms here. But we should explicitly mention we are just collecting the local shard and returning it.

pritamdamania87 commented on torch/distributed/_shard/sharding_plan/api.py:
```diff
@@ -0,0 +1,63 @@
+class ShardingPlan(object):
+    """
+    Representation of a sharding plan, describes how to shard a module
+    across hosts.
```
Maybe extend these docs a bit elaborating on `plan`, `output_plan` and `collect_local_shards` and give an example of how they might be used together.

pritamdamania87 commented on test/distributed/_shard/sharding_plan/test_sharding_plan.py:
```diff
@@ -97,6 +91,10 @@ def test_basic_sharding_plan(self):
+            local_output = local_megatron_lm(input)
+
+            # verify and make sure local and sharded output matches
+            self.assertEqual(local_output, sharded_output)
```
Probably better to have a full forward + backward + optimizer pass for local and sharded and then verify the result.

Can probably enhance this in a separate PR, but having a end to end training loop with sharder is just good validation in general.

pritamdamania87 commented on torch/distributed/_shard/api.py:
```diff
@@ -118,3 +125,142 @@ def shard_parameter(
+    process_group=None
+):
+    """
+    Shards a given module according to the provided sharding_plan. This method
```
We should probably mention in the docs that this needs to be SPMD.

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"




Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.
3. We also introduce a extensible ShardingPlanner and this extensible ShardingPlanner can be used by advanced users to build their own ShardingPlan easily.

TODO:
* implement shard_module_swapper

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

[ghstack-poisoned]

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"




Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.
3. We also introduce a extensible ShardingPlanner and this extensible ShardingPlanner can be used by advanced users to build their own ShardingPlan easily.

TODO:
* implement shard_module_swapper

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

[ghstack-poisoned]

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"




Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model, this including `plan` for sharding the parameters, and `output_plan` for tagging the output layout, `return_local_tensor` for converting back to DDP.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.

TODO:
next PR we will introduce Extensible Sharder and ShardingPlanner.

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/api.py:
```diff
@@ -223,45 +249,40 @@ def shard_module(
+                _reshard_output(mod, output_spec)
+            else:
+                raise TypeError(f"Only `ShardingSpec` is supported as output_plan for '{module_path}'")
+    # convert the output back to data parallel by calling `_collect_local_shard`
```
nit: _collect_local_shard in the docs

Commit:
Update on "[shard] Sharder and ShardingPlan prototype"




Basic ShardingPlan interface and Sharder implemention:
1. We provide `ShardingPlan` to allow user to specify all parameter sharding strategies for a given model, this including `plan` for sharding the parameters, and `output_plan` for tagging the output layout, `return_local_tensor` for converting back to DDP.
2. Introduce `shard_module` API, that could take a nn.Module, a ShardingPlan, then shard the module according to the plan.

TODO:
next PR we will introduce Extensible Sharder and ShardingPlanner.

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34695159/)!

Differential Revision: [D34695159](https://our.internmc.facebook.com/intern/diff/D34695159)

[ghstack-poisoned]
